### PR TITLE
Add an option to force parameters to be uniform

### PR DESCRIPTION
**Pull request by jashug**
**Kind:** feature

Supersedes #7439.

We support an option `Set Uniform Inductive Parameters` that abstracts inductives over their parameters before type-checking the constructors. This allows to write ex.
```
Inductive list (A : Type) := nil : list | cons : A -> list -> list.
```
This patch does not include earlier syntax which allowed for `Uniform Inductive` and `NonUniform Inductive`, as that can wait for attributes to be implemented.

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated.
- [x] Entry added in CHANGES.

### CODE REVIEW

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: Since you are defining the option at this point, you should rather write:

```
.. opt:: Uniform Inductive Parameters

   When this option is set (it is off by default)...

```

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: This paragraph is a copy from the one below. Probably a rebasing issue.

File: lib/flags.mli
Comment (Zimmi48) [inline lib/flags.mli]: @ejgallego Do you want to comment on the appropriate location for defining a new flag / option? I seem to recall that we prefer to define options at the point where they are used nowadays.

**Comment by gares** [comment]
Curiosity: the API provided by `elpi` to define inductive types uses this *very same convention*: parameters are not to be repeated (indeed they cannot vary). Conversely non-uniform parameters are given the syntax of indexes (they can vary and they must be provided), that is non-uniform parameters are uniform indexes.

File: vernac/comInductive.ml
Comment (gares) [inline vernac/comInductive.ml]: May I ask if there is any deep problem here, or it is just lack of time? @herbelin do you think it is hard to support this?

**Comment by mattam82** [comment]
Definitely like the feature and it sounds simpler than the first one involving unification. I’d like to have notations working with those as well. I suppose it’s a matter of putting the right number of underscores in the « exported » notations at each inductive occurrence. One extension option I’m immediately thinking of is to make the parameters automatically maximally implicit in the constructors, but that’s just a suggestion.

**Comment by SkySkimmer** [comment]
Would it be useful to have a syntax for non-uniform parameters when the option is set? eg
```coq
Inductive Acc {A} (R : A -> A -> Prop) | x := Acc_intro : (forall y, R y x -> Acc R y) -> Acc R x.
```
with the pipe indicating the end of uniform parameters.

**Comment by gares** [comment]
> Would it be useful to have a syntax

I'm sorry but this is off topic. The option is to no pass `R` (and not just `A`), and is orthogonal to `|` to separate parameters from non-unif parameters/indexes

File: vernac/comInductive.ml
Comment (herbelin) [reply vernac/comInductive.ml]: Hard to support notations? If the notation for the type involves at least one of the arguments that would not be mentioned as in e.g. the case of `\/`, or `/\`, the notation would stop to make sense with this at least one argument not being mentioned.

**Comment by andres-erbsen** [comment]
How does this feature compare to using a new section and Context variables for uniform parameters?

**Comment by Zimmi48** [comment]
Good question. I suppose the two are equivalent and the documentation should probably make this explicit.

**Comment by jashug** [comment]
Thank you all for the feedback.

@gares @herbelin For notations, yes, the issue is that they have to be abstracted over the parameters. I don't know how to do substitutions in constr_expr, but I suppose you could do so. I think you'll still end up with some weird edge cases like
```
Ltac get_head x := match x with ?f _ => get_head f | _ => x end.
Ltac return_head x := let x := get_head x in exact x.
Inductive Foo (n : nat) := C : # where "#" := (ltac:(return_head Foo)).
```
If my understanding is correct, this will work to define `C` with `#` meaning `Foo : Type`, but outside `#` will mean `Foo : nat -> Type`.
The issue is almost exactly the same as having notations survive sections. I suppose you could have the notation only used for defining the constructors, and not be declared globally.

@mattam82 If you use `Set Implicit Arguments. Set Contextual Implicit. Set Maximal Implicit Insertion.` then the parameters should be set maximal implicit for constructors. I can add that to the documentation if you think it would be useful.
I had intended to keep pushing for the unification approach as well. Unification can unify the parameters of the conclusion with the input parameters whether uniform or not, and this is a more principled approach than the `complete_conclusion` removed in #7254. #7254 loses the heuristic that unspecified implicit parameters (curly braces) to recursive arguments are the same as the input parameters. For non-uniform inductive parameters, that heuristic is sometimes wrong, and overrides the correct solution found by unification. This PR attempts to let users be explicit about when they are dealing with uniform parameters and when not.

@SkySkimmer The first commit introduces code to support such a split. @gares proposed a different syntax variant in #7439. If you can come up with a syntax that makes @ppedrot happy, I'm all for it (personally I'd rather have such a way of specifying both uniform and non-uniform parameters in one line (and it might be adaptable to the `Fixpoint` command as well?))

@andres-erbsen The only difference is that type inference gets the constructors and indices to work with this way, so you can say e.x. `Inductive Foo A := C : A -> Foo.` and infer that `A` is a `Type`, while `Context A.` has nothing to go on. I'll add that to the documentation.

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: The proper way to quote a command is:

```
:cmd:`Context`
```

File: lib/flags.mli
Comment (ejgallego) [reply lib/flags.mli]: Well indeed this is documented at the top of `Flags.mli`.

Please move the flags to the proper module. Also, don't pollute the namespace with `make/is` fucntions.

File: doc/sphinx/language/gallina-specification-language.rst
Comment (jashug) [reply doc/sphinx/language/gallina-specification-language.rst]: Thanks. I hadn't seen `doc/sphnix/README`, that gave me a better sense of how the new documentation works. I fixed this, and also used `undo` for the examples.

File: lib/flags.mli
Comment (jashug) [reply lib/flags.mli]: Sorry about that, I was blindly mimicking the cumulativity flag above, and didn't read `flags.mli`. I moved the option to `vernacentries.ml`, which is where it is used, but if it would be better in `ComInductive` let me know.

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: Since you use `undo`, the `Unset` command here is inessential.

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: You could perhaps add ```(cf. :ref:`section-mechanism`)``` here.

**Review by Zimmi48 — APPROVED** [review]
I approve the documentation part.

Note that there is another example of defining inductive types within sections in the "mutual inductive types" section below. Maybe this part of the doc could be made to use uniform inductive types instead.

File: vernac/vernacentries.ml
Comment (ppedrot) [inline vernac/vernacentries.ml]: Use a `()` pattern here.

File: vernac/comInductive.mli
Comment (ppedrot) [inline vernac/comInductive.mli]: Might be worth giving a label to this argument. The other ones are lacking it, but it doesn't justify not doing it systematically when introducing new arguments.

**Review by ppedrot — DISMISSED** [review]

**Comment by jashug** [comment]
Addressed changes requested by @ppederot

**Comment by jashug** [comment]
#7254 is looking good, with buy-in from @JasonGross, so I believe this PR is just waiting on approval from @mattam82. The gitlab failure of Equations is almost certainly spurious (it was green before), but I can't restart it myself.

**Comment by Zimmi48** [comment]
@jashug Yes, exit code 137 always mean spurious failure (runner system failure in GitLab terminology). I've restarted the job. You can request to join the Coq organization on GitLab https://gitlab.com/coq to be able to restart failed job.

**Comment by mattam82** [comment]
I think you’ll have to rebase first to make equations go through.

**Review by mattam82 — APPROVED** [review]

**Comment by jashug** [comment]
Rebased, all green.

**Comment by JasonGross** [comment]
Should I merge the overlay into fiat-crypto now?

**Comment by jashug** [comment]
Well, I believe that there is general agreement to move forward with this and #7254, so it seems unlikely that the overlay will change and it is backwards compatible. Do note though that the overlay is not necessary for this PR, but only for #7254.

**Pull request merged**
